### ISSUE TITLE
Expected type 'SSMClient', got 'BaseClient' instead

### ISSUE THREAD

**Issue by herzog0** [vemel/mypy_boto3_builder#40]
I didn't open a bug because I'm not sure if this is not something I'm doing wrong with the installation or usage, or that the problem isn't with PyCharm.   

The error message:  
`Expected type 'SSMClient', got 'BaseClient' instead`

The code that triggers it:  
```python
import boto3
import mypy_boto3_ssm as ssm
client: ssm.SSMClient = boto3.client("ssm") # the message is for the 'client' variable
```

The package versions:  
```
boto3==1.14.37.0
boto3-stubs==1.14.37.0

PyCharm 2020.2
```

If this should not occur, I will investigate further to know if the problem is with PyCharm's typechecks.

**Comment by takeda** [comment]
I think this bug happens to all clients. I get it with EC2, Organizations, SSM, StepFunctions.

Based on my understanding how types work I think PyCharm is right (assuming SSMClient is child of BaseClient), but mypy doesn't show this error.

**Comment by takeda** [comment]
@vlasovskikh since you're working on PyCharm, can you tell if there's something wrong with these stubs?

**Comment by tuukkamustonen** [comment]
Having the same issue with PyCharm 2020.2 (didn't try earlier) and:

```
boto3                         1.14.47
boto3-stubs                   1.14.47.0
mypy-boto3                    1.14.47.0
mypy-boto3-s3                 1.14.47.0
```

```py
import boto3
from mypy_boto3_s3 import S3Client
client: S3Client = boto3.client('s3')
```

Navigating to `boto3.client()` goes to `site-packages/mypy_boto3/boto3_init_gen.py` which shows:

```
# pylint: disable=unused-argument
def client(
    service_name: str,
    region_name: str = None,
    api_version: str = None,
    use_ssl: bool = None,
    verify: Union[str, bool] = None,
    endpoint_url: str = None,
    aws_access_key_id: str = None,
    aws_secret_access_key: str = None,
    aws_session_token: str = None,
    config: Config = None,
) -> BaseClient:
    pass
```

So it indeed returns `BaseClient` and not `S3Client`.

**Comment by vemel** [comment]
Hello guys,

Thank you for the report. I have not found any  cleaner way to annotate client classes.

As a workaround, you can run `mypy_boto3`, it will generate overloads in `site-packages/mypy_boto3/boto3_init_gen.py`.

If you have any suggestion - please let me know.

**Comment by vemel** [comment]
The latest release should fix this issue. Thank you for the report. Let me know if still does not work correctly.

**Comment by takeda** [comment]
@vemel I still see the issue on 1.14.51.1 for `ec2: EC2Client = boto3.client("ec2", region_name=region)`

Wouldn't setting `boto3.client`'s result to `Any` fix it?

**Comment by takeda** [comment]
Interestingly `ec2: EC2ServiceResource = boto3.resource("ec2", region_name=region)` is not highlighted. So anything that's different between `client()` and `resource()`?

**Comment by vemel** [comment]
THere is a difference, but I was sure I fixed both cases. Could you please try mypy on this code sample? https://github.com/vemel/mypy_boto3_builder/blob/master/examples/s3_example.py

The difference could be that `client` returns `botocore.BaseClient`, but `resource() -> Any`.

**Comment by tuukkamustonen** [comment]
@vemel Unfortunately, 1.14.53.0 still returns `BaseClient`...

I did `pip uninstall` + `pip install` (instead of `pip install -U`) to ensure there wouldn't be leftovers from previous version.

**Comment by vemel** [comment]
@tuukkamustonen there is a new branch that addresses this issue: https://github.com/vemel/mypy_boto3_builder/pull/47

Should be released in a couple of days.
